### ISSUE TITLE
EditorTest requesting many batched tests crashes the editor

### ISSUE THREAD

**Issue by Kadino** [o3de/o3de#9167]
**Describe the bug**
Tests using EditorBatchedTest crash the editor when too many batched tests are requested. This is a consistent crash.

**Steps to reproduce**
1. Open file https://github.com/o3de/o3de/blob/development/AutomatedTesting/Gem/PythonTests/Physics/TestSuite_Main.py
2. Change all tests using `EditorSharedTest` to use `EditorBatchedTest`
3. Run the tests, e.g. with `ctest -C profile -R AutomatedTesting::PhysicsTests_Main -VV`

**Expected behavior**
Multiple tests execute in serial within the same editor instance.

**Actual behavior**
Editor immediately crashes when there are more than 10-12 batched tests, with no stacktrace:
```
Gem\PythonTests\Physics\TestSuite_Main.py .FFFFFFFFxxFFF                                                                                                        [100%]

============================================================================== FAILURES ==============================================================================
__________________ TestAutomation.RigidBody_EnablingGravityWorksUsingNotificationsPoC[windows-crash_log_watchdog0-AutomatedTesting-windows_editor] ___________________
..\tools\lytesttools\ly_test_tools\o3de\editor_test.py:478: in result
    cls._report_result(test_spec.__name__, editor_test_data.results[test_spec.__name__])
..\tools\lytesttools\ly_test_tools\o3de\editor_test.py:756: in _report_result
    pytest.fail(error_str)
E   Failed: Test RigidBody_EnablingGravityWorksUsingNotificationsPoC:
E   Test CRASHED, return code 0xc0000409
E   ---------------
E   |  Stacktrace |
E   ---------------
E   -- No crash log available --
E   [Errno 2] No such file or directory: 'E:\\Amazon\\GitHub\\o3de\\AutomatedTesting\\user\\log_test_1\\error.log'------------
E   |  Output  |
E   ------------
E   -- No output --
E   --------------
E   | Editor log |
E   --------------
E   [editor_test.log]  <17:32:09> CSystem::Quit invoked from thread 48812 (main is 48812)
```

**Assets required**
N/A

**Screenshots/Video**
N/A

**Found in Branch**
Development

**Desktop/Device (please complete the following information):**
 - OS: Windows

**Additional context**
This may be related to argument parsing by the editor:
- Arguments are externally passed to the editor in this file and its siblings: https://github.com/o3de/o3de/blob/development/Tools/LyTestTools/ly_test_tools/launchers/platforms/win/launcher.py#L72
- The editor parses arguments at: https://github.com/o3de/o3de/blob/38261d08007a1ad14135cf790cb855ccca4d8595/Code/Editor/main.cpp

The in-editor Python code cannot execute in debug mode due to https://github.com/o3de/o3de/issues/4404

**Comment by moraaar** [comment]
Assigning blocking priority as this is blocking sig-core to be able to move Physics automated tests from `EditorSharedTest` to `EditorBatchedTest`.

**Comment by evanchia-ly-sdets** [comment]
Initial investigation shows that the number of batched tests do not matter, but rather the character length of the executing command. There appears to be limits both on the individual --runpythontest arg and the entire command string. Some initial ideas are:
1. Find a way to have relative filepaths to each test case module. This would most likely involve using pathlib to add each test package to the Path. This is slightly hacky and would involve some test refactor.
2. Increase the character limit of the Editor args. I'm unsure how much work this would be and if it's even feasible. With the number of tests requested to run in Batch, it would be in the magnitude of tens of thousands of characters.
3. Separate Batch runs into multiple serial batch runs based off of character breaking points. This is clean, however it would increase testing times significantly. About 10 tests can run in the same batch on average, so a test suite with 100 test cases will take a decent magnitude longer.

**Comment by Kadino** [comment]
Setting arguments in an environment variable (or modifying PYTHONPATH or PATH) may be problematic for parallel execution, but could be handled through having multiple, or only setting the values for the separate processes which are getting spun off. This would be preferable to writing/reading a file with the same info, which can hit various I/O and stale filesystem issues.

Character limits are likely being enforced by the OS shell, changing which is a brittle user-facing pain on Windows and essentially an intractable problem. It's unclear what the editor's responsibility should be, but it should probably not be crashing when a long string is provided. If we work around this, I suggest cutting a separate ticket to SIG-Core about C++ argparse crashing when args are too long.

Providing a limit (cli arg) for batch-runs seems reasonable, but doesn't actually fix this relatively narrow ceiling on how many tests can be batched together. Forcing users to set their own value OR automatically re-batching to stay under a known limit would both be a hacky workarounds.

**Comment by moraaar** [comment]
If each batch creates its own file to store all the tests arguments and the editor of that batch reads it, there shouldn't be any IO issues from running multiple editors in parallel because each editor has a different file.

1. python creates a unique file for batch and puts the tests args there.
2. python runs editor passing as argument that filename
3. editor reads the file to get all the tests it needs to run and runs all the tests
4. python deletes unique file for batch.

Would this be possible?

**Comment by Kadino** [comment]
Modifying environment variables for an executable should have the same effect as creating configuration files, and with fewer failure modes. I'm not saying writing a file will *cause* failures but that it has a greater surface area of failure.

**Comment by lumberyard-employee-dm** [comment]
> If each batch creates its own file to store all the tests arguments and the editor of that batch reads it, there shouldn't be any IO issues from running multiple editors in parallel because each editor has a different file.
> 
>     1. python creates a unique file for batch and puts the tests args there.
> 
>     2. python runs editor passing as argument that filename
> 
>     3. editor reads the file to get all the tests it needs to run and runs all the tests
> 
>     4. python deletes unique file for batch.
> 
> 
> Would this be possible?

Yes that should be possible. 
That does create an extra point of failure with the need to create a temporary file(if anything this file should be created using the temporary directory support of python).

O3DE applications actually support the [--regset-file](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Settings/SettingsRegistryMergeUtils.h#L250-L255) option which can be used to specify a path to .setreg/.setregpatch file with settings in it.

That combined with unknown feature that is actually possible to use the settings registry to store [command line options](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Settings/SettingsRegistryMergeUtils.cpp).

By using the `--regset-file` option with a `.setregpatch` file (a JSON patch formatted file) it should be possible to inject command line arguments without any changes to existing code today

### /tmp/command_args.setregpatch
```json
[
  {
  	"path": "/O3DE/Runtime/CommandLine/-",
    "op": "add",
    "value": {
      "Option": "runpythontest",
      "Value": "<arg value 1>"
    }
  },
  {
  	"path": "/O3DE/Runtime/CommandLine/-",
    "op": "add",
    "value": {
      "Option": "<arg option 2>",
      "Value": "<arg value 2>"
    }
  }
]
```
![image](https://user-images.githubusercontent.com/56135373/166067181-45f0c00c-cb31-41be-86fa-aa828b2efcd4.png)

**Comment by evanchia-ly-sdets** [comment]
If this would be the preferred fix, then we can have the editor test framework create temporary regset files and call the editor with the --regset-file option shown above.

**Comment by lumberyard-employee-dm** [comment]
> If this would be the preferred fix, then we can have the editor test framework create temporary regset files and call the editor with the --regset-file option shown above.

I wouldn't say that is a preferred fix, just one way that can work.

I think it would be better to add support for a `--command-line-file` argument where the command line arguments can be read from a file.

**Comment by evanchia-ly-sdets** [comment]
Hello, I've currently made changes to the Editor batched framework to create a temp file with the test modules listed in it and changing Editor behavior to accept a file to read. Testing everything at the moment and will release a PR once it's ready. Will definitely add you as a reviewer once that's done.

**Comment by evanchia-ly-sdets** [comment]
https://github.com/o3de/o3de/pull/9463

**Comment by hultonha** [comment]
> Hello, I've currently made changes to the Editor batched framework to create a temp file with the test modules listed in it and changing Editor behavior to accept a file to read. Testing everything at the moment and will release a PR once it's ready. Will definitely add you as a reviewer once that's done.

Hey @evanchia-ly-sdets, that's great to hear, thanks for letting me know. I just had a look at the PR and left some feedback, thanks for your help with this! 🙂

**Comment by AlexOteiza** [comment]
Hi,

I always suspected that this would lead into a problem at some point.. 
I would be a bit against having an IO dependency, as this proved to cause problems like race conditions (cc/ @amznestebanpapp).

I feel like it would be more interesting to have a cmdline argument so the editor can read the tests names via stdin, (like many linux applications). something like --runpythontests_stdin. 
For example the linux program `cat` reads from stdin if we don't provide a filename as argument. This means that a python program can just run `cat` and send via stdin the filename to read. 

LyTestTools already creates a pipe (reading stdout) when running O3DE editor. So its just a matter of sending data through stdin, which will be very easy to do as the setup is already in place

**Comment by AlexOteiza** [comment]
> > I would be a bit against having an IO dependency, as this proved to cause problems like race conditions
> 
> Each batch execution it's using a different file, why would that cause a race condition?

A race condition would happen if the file is not ready for the editor to read. This happened in the past when reading log files (in the reverse direction)

**Comment by AlexOteiza** [comment]
> > A race condition would happen if the file is not ready for the editor to read. This happened in the past when reading log files (in the reverse direction)
> 
> The file (temporary file is used) is being written and closed before editor is launched, so there should be no problem.

The editor could not see the file if there is not enough time between the creation and the launch, this is exactly what happened in the past. Its doable, but it adds possible points of failure that proved to be an issue in the past as I mentioned.

Reading from stdin is safer, simpler and would lead to less code complexity, not needing to manage multiple and temporary files for different editors.

**Comment by moraaar** [comment]
Fixed by #9578 9167
